cpp/src/arrow/engine/substrait/util.cc

```diff
@@ -0,0 +1,129 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/engine/substrait/util.h"
+#include "arrow/util/async_generator.h"
+#include "arrow/util/async_util.h"
+
+namespace arrow {
+
+namespace engine {
+
+namespace substrait {
+
+namespace {
+/// \brief A SinkNodeConsumer specialized to output ExecBatches via PushGenerator
+class SubstraitSinkConsumer : public compute::SinkNodeConsumer {
```

Comment:
This seems useful more generally than just for Substrait, and it doesn't depend on Substrait, so consider moving this to `arrow/engine`.

Reply:
If you are running outside of substrait then `SinkNode` provides exactly this behavior.  This type is only needed because Substrait forces you into a `ConsumingSInkNode` so unless other places are doing this I'm not sure it would be generally useful.